src/Access/ContextAccess.cpp

```diff
@@ -318,8 +318,12 @@ void ContextAccess::setRolesInfo(const std::shared_ptr<const EnabledRolesInfo> &
     roles_info = roles_info_;
     enabled_row_policies = access_control->getEnabledRowPolicies(
         *params.user_id, roles_info->enabled_roles);
-    enabled_quota = access_control->getEnabledQuota(
-        *params.user_id, user_name, roles_info->enabled_roles, params.address, params.forwarded_address, params.quota_key);
+
+    /// TCP connection handler receives quota key after session creation
```

Comment:
I don't really understand the changes around the `quota_key`. How is it related to the session tracking?

Reply:
I'm not sure if my solution is the best, but I haven't found a better and simpler solution to not change the current TCP protocol and make both client and server work without errors.
If you have better ideas, I will be happy to implement them.

During a session between client and server, the client assumes that there should be no errors on the server side after the server has replied Hello.
However, with the addition of the `max_sessions_per_user` constraint, throwing an error after `server hello`  is now possible. To get `max_sessions_per_user `we need to reach the `profile settings`, and to do that we need to create a `session context`.
Creating a session context and creating a quota without a quota key can throw exception.

To better understand the problem, I have created attached images. 
I hope they will make it easier to understand.

![bad_variant_make_session_context_after_sending_hello](https://github.com/ClickHouse/ClickHouse/assets/16054347/d48163c6-bca9-4518-89b5-c01cf90e766b)
![bad_variant_quota_key_initialize](https://github.com/ClickHouse/ClickHouse/assets/16054347/7f7b189a-c1e7-4a59-af37-b941ebb5312d)
![good_variant](https://github.com/ClickHouse/ClickHouse/assets/16054347/73ec555f-7276-431e-8576-30d3a71e142e)

Reply:
Thanks for the detailed explanation and the figures, I understood the problem now 😄

I am not an expert with the TCP protocol, but the current solution looks fine to me (not great but also not terrible) - we can go with it. My main question at this point is about this assumption:

> During a session between client and server, the client assumes that there should be no errors on the server side after the server has replied Hello.

That's interesting. What happens exactly on the client? Does the connection abruptly end?

I see that there is a handler for exceptions: `sendException()`. Just curious, would it be possible to use this to send an exception after "Hello" to the client?

Reply:
>That's interesting. What happens exactly on the client? Does the connection abruptly end?

Server closes connection.
If exception happens after hello received from server client tries to send quota key if it has it. Usually server doesn't have time to close the connection while the client is sending quota key, this step passes (Client and server are on the same computer).
If the client was in interactive mode then it waits for query to enter.
After client has an query then it tries to send it to the server. Client detects that connection is closed by Connection::ping and doesn't print any error (client does not read exception code at all), and finally client tries to reconnect and reauthenticate.

> I see that there is a handler for exceptions: sendException(). Just curious, would it be possible to use this to send an exception > after "Hello" to the client?

Yes, but client doesn't read the socket data after receiving hello from server, because it does not expect an error.